sql/sqlwire/wire.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sqlwire
+
+import "strconv"
+
+func (d Datum) String() string {
+	if d.BoolVal != nil {
+		if *d.BoolVal {
+			return "true"
+		}
+		return "false"
+	}
+	if d.IntVal != nil {
+		return strconv.FormatInt(*d.IntVal, 10)
+	}
+	if d.UintVal != nil {
+		return strconv.FormatUint(*d.UintVal, 10)
+	}
+	if d.FloatVal != nil {
+		return strconv.FormatFloat(*d.FloatVal, 'g', -1, 64)
+	}
+	if d.BytesVal != nil {
+		return string(d.BytesVal)
+	}
+	if d.StringVal != nil {
+		return *d.StringVal
+	}
+	return "NULL"
+}
+
+// IsNull returns true iff the datum is NULL.
+func (d Datum) IsNull() bool {
+	return d.BoolVal == nil && d.IntVal == nil && d.UintVal == nil &&
+		d.FloatVal == nil && d.BytesVal == nil && d.StringVal == nil
+}
+
+// Bool parses the datum as a bool.
+func (d Datum) Bool() bool {
+	if d.BoolVal != nil {
+		return *d.BoolVal
+	}
+	if d.IntVal != nil {
+		return *d.IntVal != 0
+	}
+	if d.UintVal != nil {
+		return *d.UintVal != 0
+	}
+	if d.FloatVal != nil {
+		return *d.FloatVal != 0
+	}
+	if d.BytesVal != nil {
+		v, err := strconv.ParseBool(string(d.BytesVal))
+		if err != nil {
+			return false
+		}
+		return v
+	}
+	if d.StringVal != nil {
+		v, err := strconv.ParseBool(*d.StringVal)
```

Comment:
Mainly I'd prefer to have a lot fewer implicit conversions; that approach has served Go well. To the extent that we need conversions we should avoid "skipping levels" in the conversion graph. For example, instead of using `ParseBool` which partially overlaps with the numeric parsers (but only for the special values 1 and 0), we should instead look for the words "true" and "false", and if we don't have a match try going through a pipeline like `string -> number -> bool` instead of introducing special cases that are only available for the case of a direct `string -> bool` conversion.